src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,150 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// --------------------------------------------------------------------------------
+// NativeImage.cpp
+//
+
+// --------------------------------------------------------------------------------
+
+#include "common.h"
+#include "nativeimage.h"
+
+// --------------------------------------------------------------------------------
+// Headers
+// --------------------------------------------------------------------------------
+
+#include <shlwapi.h>
+
+AssemblyNameIndex::AssemblyNameIndex(const SString& name, int32_t index)
+    : Name(name), Index(index)
+{
+    Name.Normalize();
+}
+
+NativeImage::NativeImage(
+    PEFile *pPeFile,
+    PEImageLayout *pPeImageLayout,
+    READYTORUN_HEADER *pHeader,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker& amTracker)
+    : m_eagerFixupsLock(CrstNativeImageEagerFixups)
+{
+    CONTRACTL
+    {
+        CONSTRUCTOR_CHECK;
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    m_utf8SimpleName = nativeImageName;
+    m_pPeImageLayout = pPeImageLayout;
+    m_eagerFixupsHaveRun = false;
+    
+    m_pReadyToRunInfo.Assign(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(ReadyToRunInfo))))
```

Comment:
This is allocated on LoaderHeap, but stored in NewHolder. NewHolder is for things allocated using regular `new`. Things allocated on LoaderHeap should not be stored in NewHolder.

It may be best to use regular new for everything related to this, and not use LoaderHeap at all.

Reply:
Fixed in 14th commit.